### ISSUE TITLE
[Gradle Plugin] Option to make detekt system resource availability aware

### ISSUE THREAD

**Issue by plastiv** [arturbosch/detekt#2443]
Your feature request may already be reported!
Please search in the [issues](https://github.com/arturbosch/detekt/issues) before creating a new one.

## Expected Behavior

Configuration option to control amount of detekt processes scheduled in parallel. Similar to `maxFork` option from Gradle work api. 

## Current Behavior

Having multi-module project with 50 modules. Configured like this:

```groovy
subprojects { subProject ->
            subProject.apply plugin: 'io.gitlab.arturbosch.detekt'

            detekt {
                debug = false
                toolVersion = versions.detektPlugin
                input = subProject.files("src")
                config = files("$rootDir/code_quality_tools/detekt-config.yml")

                reports {
                    xml.enabled = false
                    txt.enabled = false
                    html.enabled = true
                }
            }
        }
    }
}
```

Running `./gradlew detekt` results in (a try to spawn) 50 java processes at the same time. Which hits the OOM wall on a CI containirezed machine.

Practically, we have to disable running detekt in parallel and run detekt module-after-module.

Checking every module in a multi-module project looks like unsupported use case for now? 

## Context

Related issues I found: 
* https://github.com/arturbosch/detekt/issues/1295
* https://github.com/arturbosch/detekt/issues/2035
* https://github.com/arturbosch/detekt/issues/2119

**Comment by schalkms** [comment]
Thanks for the feedback and for describing your use case in detail!
I can understand the limitations in some environments and how it hinders further adoption of detekt.
I'm not sure whether the value/effort for supporting this use case is given here. @arturbosch might want to assess first.

**Comment by plastiv** [comment]
For the past 2-3 years android community shifts into modularizing the app into pieces approach. Its movement has been recognized by Google dev team. And these days Google dev relation team is encoraging application development by splitting its code into separate gradle projects as well. https://www.youtube.com/watch?v=PZBg5DIzNww

My understanding that the same target audience that is interested in having healthy code base and having clean code by setting up static anylizers is also the same audience which tends to follow suggested best practices. 

And so I think projects which install detekt are likely to be projects organized with submodules. And supporting multi-module case is not that uncommon. But of course you as maintainer knows better your core supported use case.

Would you recommend to drop detekt gradle pluging and switch to execute `jar` directly? Possibly with some custom helper scripts on top for multi-module projects?

**Comment by arturbosch** [comment]
@plastiv starting from detekt 1.7.0 we won't spawn any java processes anymore as we have moved to a cached classloader approach.
Please test `1.7.0-beta2` on your multi module project and share your experience.
Note that this also boosts performance by a factor of 2-4.

**Comment by plastiv** [comment]
> If you use org.gradle.parallel=true gradle should be processor aware and only start #cores of detekt tasks?

Correct. I am using a CI machine with 64 cores. If I limit parallel option explicitly `org.gradle.workers.max=8` OOM is still reproducable for me. Which was one of the reasons I have created this issue, since other tasks (kotlin compile, android assemble, ktlint, checkstyle) seems to have no issues running in 64 workers. 

So it looks like my issue is triggered by detekt. But now I'm not sure. I have learned recently that `org.gradle.jvmargs=` isn't the total build memory budget. Instead `org.gradle.jvmargs=` controls gradle daemon only. Kotlin daemon, jvm test forks and any external gradle plugin code (like detekt) is free to ignore and use it's own settings. Which makes sense, of course. But it limits control options to profile and monitor total memory consumption per CI run. 

I have tried but so far I couldn't have consistent results with `1.7.0-beta2` either. Mostly due to lack of profiling tool. There is a noise in build output triggered by other changes. Maybe you can suggest what to use to fine tune memory settings? 

I also plan to make more investigation next week and will report my findings.

**Comment by arturbosch** [comment]
@plastiv how large is your codebase? 

I know of a specific "memory leak" in detekt which got deprecated in 1.7.0 and needs to wait until 1.8.0 to get removed. It basically does not free memory (from the Config objects) until all files are analyzed.

> I have tried but so far I couldn't have consistent results with 1.7.0-beta2 either.
Does it mean that sometimes still OOM occur?

I guess if you have 64 cores you have increased the gradle daemon memory to minimum of 8gb?
What kind of OOM occurs? Metaspace OOM maybe (with 1.7.0-beta2)?

**Comment by plastiv** [comment]
I've invested more into my issue research and I think now that I have misinterpreted original error message

```java
 java.lang.OutOfMemoryError: unable to create new native thread
 	at java.lang.Thread.start0(Native Method)
 	at java.lang.Thread.start(Unknown Source)
 	at java.util.concurrent.ThreadPoolExecutor.addWorker(Unknown Source)
 	at java.util.concurrent.ThreadPoolExecutor.execute(Unknown Source)
```

as an error triggered by exhausting memory budget. Which I later connected with known classpath issue. 

Turns out the same exception is thrown when reaching out PIDs max limit. Enforcing PID limit is (new) feature on kubernetes https://kubernetes.io/blog/2019/04/15/process-id-limiting-for-stability-improvements-in-kubernetes-1.14/ which is managing our CI nodes. I can confirm that build is capable to start 500 threads per run by checking VisualVM locally. 

I do still think that spawning over 500 threads with `org.gradle.workers.max=8` is smelly. But I'm almost certain that it is a coincidence that thread # 501 happened to be a detekt's thread. **My issue is outside of the detekt project.** 

Looking forward to the days when android projects could be built on JDK11. Which has Java flight recorder built in and which is container aware out of the box. This would save my investigation time considerably.

> how large is your codebase?

2.5k files, 100k lines of code, 60% kotlin, 40% java

> Does it mean that sometimes still OOM occur?

Yes. But I currently think that OOM swallows out of PID exception. Looks like it was not considered as important detail to specify. 

> I guess if you have 64 cores you have increased the gradle daemon memory to minimum of 8gb?

Build peaks 2.7Gb only for gradle daemon, 1.2Gb for kotlin daemon and if thread stack size is 8Mb (apparently, default these days) then 500 threads == 4Gb just for the stacks (hello kotlin coroutines 👋where are you?)

> Metaspace OOM maybe

Gradle daemon peaks 300Mb metaspace and Kotlin daemon 100Mb. Unless specified metaspace has no limit. But if reached have clear error message that metaspace is the issue (checked before).

**Comment by schalkms** [comment]
Thanks for the detailed feedback and analysis @plastiv. It's always interesting to see how detekt is used in different scenarios.
I guess this can be closed now, since there's (atm) nothing we can do about it, right?

**Comment by arturbosch** [comment]
@plastiv thanks for sharing your profiling experience.
I find it interesting and challenging to profile applications.
Lastly I hunted some classloader memory leaks in detekt with good old VisualVM :).
